### PR TITLE
DS3 Linux fix

### PR DESCRIPTION
**Pull request by RipleyTom**
Fix for the ds3 pad handler for linux.
Thanks to z0z0z for testing.

### CODE REVIEW

**Comment by z0z0z** [comment]
Didn't work for me, even with a udev rule. Thank you though.

**Comment by RipleyTom** [comment]
It needs retesting, reimplemented all the usb interactions through hid api(it wouldn't work on windows but it might work on linux).

**Comment by z0z0z** [comment]
Getting somewhere. Very impressive.

With a PS3 controller connected through USB, if I open the pad settings for native DS3 handler, it triggers warn_about_drivers and I see this in console

`E HLE: [DS3] One or more DS3 pads were detected but couldn't be interacted with directly`

Here's my udev rules in `/etc/udev/rules.d/98-ds3.rules`

`KERNEL=="hidraw*", ATTRS{idVendor}=="054c", ATTRS{idProduct}=="0268", MODE="0660", TAG+="uaccess"`

lsusb

`Bus 003 Device 010: ID 054c:0268 Sony Corp. Batoh Device / PlayStation 3 Controller`

**Comment by z0z0z** [comment]
Maybe DS3 controllers don't work with hidapi? I made a stripped down version of this https://github.com/signal11/hidapi/blob/master/hidtest/hidtest.cpp

```
#include <stdio.h>
#include <hidapi.h>

int main()
{
        hid_device *handle;

	// Open the device using the VID, PID,
	// and optionally the Serial number.
	////handle = hid_open(0x4d8, 0x3f, L"12345");

	handle = hid_open(0x054C, 0x5C4, NULL);
	if (!handle) {
		printf("unable to open device\n");
 		return 1;
	}

	hid_close(handle);
	hid_exit();

	return 0;
}
```

Compiled with `g++ -lhidapi-hidraw -I/usr/include/hidapi ds3simple.cpp`

This is the PS4 controller VID and PID. It works with my PS4 controller, I don't get any message when running this and having a PS4 controller connected through USB.

I tried `0x0268` and `0x268` as the PID, but this program will print `unable to open device` if I run it with a PS3 controller connected through USB.

**Comment by RipleyTom** [comment]
Technically ds3 and ds4 use the same drivers on linux so there is no reason it shouldn't work.

**Comment by z0z0z** [comment]
> Technically ds3 and ds4 use the same drivers on linux so there is no reason it shouldn't work.

I know they both use hid-sony kernel module

But for example I know when connecting through bluetooth, the PS3 controller actually requires a special Bluez plugin in userspace, and an extra authorization prompt, whereas PS4 controllers through bluetooth don't need a plugin, or authorization prompt.

https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/plugins/sixaxis.c

You would know much more than me on what a PS3 controller does. I found this github issue with a response from the hidapi developer on PS3 controllers https://github.com/signal11/hidapi/issues/74

Btw you can compile hidapi programs with `-lhidapi-hidraw` or `-lhidapi-libusb`, libusb didn't work.

If you have any suggestions or code you want me to test for Linux I'll always be happy to.

**Comment by RipleyTom** [comment]
Yes BT part is really funky(which is why I didn't implement it) but wired part should be similar to DS4 and I'm assuming you try to use it wired, right?

**Comment by z0z0z** [comment]
> Yes BT part is really funky(which is why I didn't implement it) but wired part should be similar to DS4 and I'm assuming you try to use it wired, right?

Yeah all my tests have been wired.

**Comment by z0z0z** [comment]
Ignore all my stupid rambling.

My udev rule had `MODE="0660"` instead of `MODE="0666"`.

This actually makes my PS3 controller work fine with the native DS3 handler in RPCS3 now.

**Comment by z0z0z** [comment]
Sixaxis motion stuff working as of latest commit.

Bluetooth also works, but requires a seperate udev rule, it won't work without it I've tried. So this is what my rules file looks like 

```
# DS3 Bluetooth                    
KERNEL=="hidraw*", KERNELS=="*054C:0268*", MODE="0666"

# DS3 USB                    
KERNEL=="hidraw*", ATTRS{idVendor}=="054c", ATTRS{idProduct}=="0268", MODE="0666"
```

File: rpcs3/ds3_pad_handler.cpp [hunk 1]
```diff
@@ -25,30 +25,55 @@ ds3_pad_handler::~ds3_pad_handler()
...
+#endif
+	return true;
+}
+
+bool ds3_pad_handler::Init()
+{
+	if (is_init)
+		return true;
+
+	if (!init_usb()) return false;
```
Comment (AniLeo) [inline rpcs3/ds3_pad_handler.cpp]: newline

**Comment by RipleyTom** [comment]
Added the udev rules to the error message if it fails to open the device with hidapi on linux.

**Comment by mirh** [comment]
Wait a minute, are you saying this also work for pressure sensitivity?

**Comment by RipleyTom** [comment]
Ofc, you can test it with https://www.psx-place.com/resources/gamepad-test.498/

**Comment by mirh** [comment]
:thinking: 
Ehrm.. Are you sure you aren't using kernel 4.14 or earlier?
Because I have a very hard time believing programs compiled 3 years ago are still working the same.

**Comment by RipleyTom** [comment]
You realize you're talking on a PR of a ps3 emulator and I linked you a ps3 homebrew program to run on said emulator? It's an easy way to see if pad handler works properly.
Anyway this is a closed PR, come to discord if you want to discuss it more.

**Pull request merged**